File: java/src/main/java/com/miniraft/Cluster.java
// A deterministic in-memory cluster that drives time and delivers messages.
//
// There are no real timers or sockets. step() advances every node one step and
// routes whatever messages they emit. Message delivery order is stable, and the
// network can be partitioned or healed, so an entire election or a replication
// round replays identically every run - which is what makes the tests reliable.

package com.miniraft;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Deque;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Random;
import java.util.Set;

public final class Cluster {

    private final Map<Integer, RaftNode> nodes = new LinkedHashMap<>();
    private int time = 0;
    private int dropped = 0;

    private final Random rng;
    private final double dropProb;
    private final double duplicateProb;
    private final boolean reorder;

    private final Deque<Message> inbox = new ArrayDeque<>();
    // A pair (a, b) in here means the link a<->b is cut in both directions.
    private final Set<Long> partitioned = new HashSet<>();

    public Cluster(int size, int electionTimeout, int heartbeatInterval,
                   Long seed, double dropProb, double duplicateProb, boolean reorder) {
        List<Integer> ids = new ArrayList<>();
        for (int i = 0; i < size; i++) ids.add(i);
        this.rng = seed != null ? new Random(seed) : null;
        // Stagger election timeouts by node id. Real Raft randomizes these so the
        // cluster doesn't split the vote forever; a deterministic per-node offset
        // gives the same effect while keeping the simulation reproducible. When a
        // seed is given we randomize timeouts in [T, 2T) like the paper recommends.
        for (int i : ids) {
            int timeout = rng != null
                    ? electionTimeout + rng.nextInt(electionTimeout) // [T, 2T-1]
                    : electionTimeout + i * 2;
            nodes.put(i, new RaftNode(i, ids, timeout, heartbeatInterval));
        }
        this.dropProb = dropProb;
        this.duplicateProb = duplicateProb;
        this.reorder = reorder;
    }

    public Cluster(int size) {
        this(size, 10, 3, null, 0.0, 0.0, false);
    }

    public Cluster(int size, int electionTimeout, int heartbeatInterval) {
        this(size, electionTimeout, heartbeatInterval, null, 0.0, 0.0, false);
    }

    // ---- network control ----

    private static long key(int a, int b) {
        int lo = Math.min(a, b), hi = Math.max(a, b);
        return ((long) lo << 32) | (hi & 0xffffffffL);
    }

    public void partition(int a, int b) { partitioned.add(key(a, b)); }
    public void heal(int a, int b) { partitioned.remove(key(a, b)); }
    public void healAll() { partitioned.clear(); }

    public void isolate(int node) {
        for (int other : nodes.keySet()) if (other != node) partition(node, other);
    }

    private boolean reachable(int src, int dst) {
        return !partitioned.contains(key(src, dst));
    }

    // ---- driving time ----

    public void step() {
        time++;

        List<Message> pending = new ArrayList<>(inbox);
        inbox.clear();

        if (reorder && rng != null && pending.size() > 1) {
            Collections.shuffle(pending, rng);
        }

        for (Message msg : pending) {
            if (!(reachable(msg.src(), msg.dst()) && nodes.containsKey(msg.dst()))) continue;
            if (rng != null && dropProb > 0 && rng.nextDouble() < dropProb) {
                dropped++;
                continue;
            }
            int deliveries = 1;
            if (rng != null && duplicateProb > 0 && rng.nextDouble() < duplicateProb) {
                deliveries = 2; // deliver the same message twice
            }
            for (int d = 0; d < deliveries; d++) {
                enqueue(nodes.get(msg.dst()).handle(msg));
            }
        }

        for (RaftNode node : nodes.values()) {
            enqueue(node.tick());
        }
    }

    public void run(int steps) {
        for (int i = 0; i < steps; i++) step();
    }

    public RaftNode runUntilLeader(int maxSteps) {
        for (int i = 0; i < maxSteps; i++) {
            step();
            RaftNode leader = leader();
            if (leader != null) return leader;
        }
        return null;
    }

    public RaftNode runUntilLeader() {
        return runUntilLeader(200);
    }

    private void enqueue(List<Message> messages) {
        for (Message m : messages) if (reachable(m.src(), m.dst())) inbox.add(m);
    }

    // ---- inspection ----

    public Map<Integer, RaftNode> nodes() { return nodes; }
    public int time() { return time; }
    public int dropped() { return dropped; }

    /** The unique current-term leader, or null if there's a split/no leader. */
    public RaftNode leader() {
        List<RaftNode> leaders = leaders();
        if (leaders.isEmpty()) return null;
        int topTerm = leaders.stream().mapToInt(RaftNode::currentTerm).max().orElse(0);
        List<RaftNode> top = new ArrayList<>();
        for (RaftNode n : leaders) if (n.currentTerm() == topTerm) top.add(n);
        return top.size() == 1 ? top.get(0) : null;
    }

    public List<RaftNode> leaders() {
        List<RaftNode> out = new ArrayList<>();
        for (RaftNode n : nodes.values()) if (n.role() == RaftNode.Role.LEADER) out.add(n);
        return out;
    }

    public List<Object> committedCommands(int nodeId) {
        List<Object> out = new ArrayList<>();
        for (Log.Entry e : nodes.get(nodeId).applied()) out.add(e.command());
        return out;
    }
}

File: csharp/Cluster.cs
// A deterministic in-memory cluster that drives time and delivers messages.
//
// There are no real timers or sockets. Step() advances every node one step and
// routes whatever messages they emit. Message delivery order is stable, and the
// network can be partitioned or healed, so an entire election or a replication
// round replays identically every run - which is what makes the tests reliable.

namespace Coracle;

public sealed class Cluster
{
    public Dictionary<int, RaftNode> Nodes { get; } = new();
    public int Time { get; private set; }
    public int Dropped { get; private set; } // count of messages the network threw away

    private readonly Random? _rng;
    private readonly double _dropProb;
    private readonly double _duplicateProb;
    private readonly bool _reorder;

    private readonly Queue<Message> _inbox = new();
    // A pair (a, b) in here means the link a<->b is cut in both directions.
    private readonly HashSet<(int, int)> _partitioned = new();

    public Cluster(int size, int electionTimeout = 10, int heartbeatInterval = 3,
        int? seed = null, double dropProb = 0.0, double duplicateProb = 0.0, bool reorder = false)
    {
        var ids = Enumerable.Range(0, size).ToList();
        _rng = seed is not null ? new Random(seed.Value) : null;
        // Stagger election timeouts by node id. Real Raft randomizes these so the
        // cluster doesn't split the vote forever; a deterministic per-node offset
        // gives the same effect while keeping the simulation reproducible. When a
        // seed is given we randomize timeouts in [T, 2T) like the paper recommends.
        foreach (var i in ids)
        {
            int timeout = _rng is not null
                ? _rng.Next(electionTimeout, 2 * electionTimeout) // [T, 2T-1]
                : electionTimeout + i * 2;
            Nodes[i] = new RaftNode(i, ids, timeout, heartbeatInterval);
        }

        _dropProb = dropProb;
        _duplicateProb = duplicateProb;
        _reorder = reorder;
    }

    // ---- network control ----

    private static (int, int) Key(int a, int b) => a < b ? (a, b) : (b, a);

    public void Partition(int a, int b) => _partitioned.Add(Key(a, b));
    public void Heal(int a, int b) => _partitioned.Remove(Key(a, b));
    public void HealAll() => _partitioned.Clear();

    public void Isolate(int node)
    {
        foreach (var other in Nodes.Keys)
            if (other != node) Partition(node, other);
    }

    private bool Reachable(int src, int dst) => !_partitioned.Contains(Key(src, dst));

    // ---- driving time ----

    public void Step()
    {
        Time++;

        var pending = new List<Message>(_inbox);
        _inbox.Clear();

        if (_reorder && _rng is not null && pending.Count > 1)
            Shuffle(pending, _rng);

        foreach (var msg in pending)
        {
            if (!(Reachable(msg.Src, msg.Dst) && Nodes.ContainsKey(msg.Dst))) continue;
            if (_rng is not null && _dropProb > 0 && _rng.NextDouble() < _dropProb)
            {
                Dropped++;
                continue;
            }
            int deliveries = 1;
            if (_rng is not null && _duplicateProb > 0 && _rng.NextDouble() < _duplicateProb)
                deliveries = 2; // deliver the same message twice
            for (int d = 0; d < deliveries; d++)
                Enqueue(Nodes[msg.Dst].Handle(msg));
        }

        foreach (var node in Nodes.Values)
            Enqueue(node.Tick());
    }

    public void Run(int steps)
    {
        for (int i = 0; i < steps; i++) Step();
    }

    public RaftNode? RunUntilLeader(int maxSteps = 200)
    {
        for (int i = 0; i < maxSteps; i++)
        {
            Step();
            var leader = Leader();
            if (leader is not null) return leader;
        }
        return null;
    }

    private void Enqueue(List<Message> messages)
    {
        foreach (var m in messages)
            if (Reachable(m.Src, m.Dst)) _inbox.Enqueue(m);
    }

    // Fisher-Yates using the same RNG, matching Python's random.shuffle semantics
    // (any uniform shuffle is fine - the chaos tests assert invariants, not order).
    private static void Shuffle(List<Message> list, Random rng)
    {
        for (int i = list.Count - 1; i > 0; i--)
        {
            int j = rng.Next(i + 1);
            (list[i], list[j]) = (list[j], list[i]);
        }
    }

    // ---- inspection ----

    /// <summary>The unique current-term leader, or null if there's a split/no leader.</summary>
    public RaftNode? Leader()
    {
        var leaders = Nodes.Values.Where(n => n.Role == Role.Leader).ToList();
        if (leaders.Count == 0) return null;
        int topTerm = leaders.Max(n => n.CurrentTerm);
        var top = leaders.Where(n => n.CurrentTerm == topTerm).ToList();
        return top.Count == 1 ? top[0] : null;
    }

    public List<RaftNode> Leaders() => Nodes.Values.Where(n => n.Role == Role.Leader).ToList();

    public List<object> CommittedCommands(int nodeId) =>
        Nodes[nodeId].Applied.Select(e => e.Command).ToList();
}
